src/core/server/saved_objects/migrationsv2/actions/index.ts

```diff
@@ -0,0 +1,787 @@
+      // Update documents in batches of 100 documents at a time
+      // TODO: profile performance to see how much difference `scroll_size` makes
+      scroll_size: BATCH_SIZE,
+      // force a refresh so that we can query the updated index immediately
+      // after the operation completes
+      refresh: true,
+      // Create a task and return task id instead of blocking until complete
+      wait_for_completion: false,
+    })
+    .then(({ body: { task: taskId } }) => {
+      return Either.right({ taskId });
+    })
+    .catch(catchRetryableEsClientErrors);
+};
+
+export interface ReindexResponse {
+  taskId: string;
+}
+
+/**
+ * Reindex documents from the `sourceIndex` into the `targetIndex`. Returns a
+ * task ID which can be tracked for progress.
+ *
+ * @remarks This action is idempotent allowing several Kibana instances to run
+ * this in parralel. By using `op_type: 'create', conflicts: 'proceed'` there
+ * will be only one write per reindexed document.
+ *
+ * @param client
+ * @param sourceIndex
+ * @param targetIndex
+ * @param reindexScript
+ */
+export const reindex = (
+  client: ElasticsearchClient,
+  sourceIndex: string,
+  targetIndex: string,
+  reindexScript: Option.Option<string>
+): TaskEither.TaskEither<RetryableEsClientError, ReindexResponse> => () => {
+  return client
+    .reindex({
+      // Require targetIndex to be an alias. Prevents a new index from being
+      // created if targetIndex doesn't exist.
+      body: {
+        // Ignore version conflicts from existing documents
+        conflicts: 'proceed',
+        source: {
+          index: sourceIndex,
+          // Set batch size
+          size: BATCH_SIZE,
+        },
+        dest: {
+          index: targetIndex,
+          // Don't override existing documents, only create if missing
+          op_type: 'create',
+        },
+        script: Option.fold(
+          () => undefined,
+          () => ({
+            source: reindexScript,
+            lang: 'painless',
+          })
+        )(reindexScript),
+      },
+      // force a refresh so that we can query the target index
+      refresh: true,
+      // Create a task and return task id instead of blocking until complete
+      wait_for_completion: false,
+    })
+    .then(({ body: { task: taskId } }) => {
+      return Either.right({ taskId });
+    })
+    .catch(catchRetryableEsClientErrors);
+};
+
+export const waitForReindexTask = flow(
+  waitForTask,
+  TaskEither.chain(
+    (
+      res
+    ): TaskEither.TaskEither<
+      | { type: 'index_not_found_exception'; index: string }
+      | { type: 'target_index_had_write_block' }
+      | RetryableEsClientError,
+      'reindex_succeeded'
+    > => {
+      const failureIsAWriteBlock = ({
+        cause: { type, reason },
+      }: {
+        cause: { type: string; reason: string };
+      }) =>
+        type === 'cluster_block_exception' &&
+        reason.match(/index \[.+] blocked by: \[FORBIDDEN\/8\/index write \(api\)\]/);
```

Comment:
The `target_index_had_write_block` is the only failure we expect that can happen even if the algorithm is running successfully. Other failures could occur but we have no way to recover from them other than retrying. I think it's "dangerous" to go into an infinite retry loop for an error we don't understand, so in such a case we rather throw and crash the Kibana process, then at least we can start the algorithm afresh and make sure we're running actions based on the latest state in the index.

So if ES runs out of disk space it will automatically add a read-only / allow delete block and we'll start getting failures with `FORBIDDEN/12/index read-only / allow delete (api)`. But these will just be thrown and crash Kibana.